Sara/mkt 293 document workbencheditor (#432)
returntocorp/semgrep-docs: docs/semgrep-app/editor.md

The Editor is free to use on all tiers, but saving a rule for **private use** is a **Team/Enterprise tier feature**.

## Accessing Semgrep Editor

1. Sign in to your [Semgrep App account](https://semgrep.dev/login).
2. Click **Editor (beta)** from the **App sidebar**. The **splash page** and **library pane** appear.
3. Do any of the following steps:
    1. To create a new rule, click on the **(+) plus sign** or **Create new rule** button.
    2. To open any rule you’ve recently edited, select it from the **Recent** list.
    3. To view a sample rule, select it from the **Examples** list. The rule renders within the Editor.
    4. To start a tutorial or read the docs, select it from the **Learn** list. This navigates you away from the Editor.

## Viewing rules

> ![Screenshot of the Editor's main view composed of three panes and top menu](../img/editor-panes.png)

Reviewer: Could I send you a different screenshot of this view? The Editor's splash page looks best if there are multiple rules in the "Recent" column (something like this)

<img width="1595" alt="Screen Shot 2022-02-08 at 9 04 44 AM" src="https://user-images.githubusercontent.com/19784302/153038117-280f11c3-5084-4c70-bb24-7aeb4dbffd69.png">

Author: I will use this photo you provided, thank you!

The editor is composed of three panes and a top menu.

<dl>

>     <dt>Library</dt>

Reviewer: I'm finding this section a bit hard to read and I think it's because of the combination of indented lines and bullet points). Is there a convention here? Is it possible to standardize on either indent or bullet points to start a line for consistency?

Author: Hmm.. I agree. this is a CSS-ish design issue. (As with the borders that @pabloest pointed out - the images are too closely spaced to the text, too, but that will take a separate PR.) The dt/dd elements are not sufficiently grouped together and the ul/lis (within the dd) have their own spacing, resulting in an inconsistent look and lack of grouping. I'll try to think of a quick solution for now.

        </ul>
        <strong>Private rules</strong> are denoted by a <strong>lock</strong> icon overlaid over the file icon.
    </dd>
    <dt>YAML editor</dt>
    <dd>Enter your rule’s YAML schema in this pane.</dd>
    <dt>Sample code window</dt>
    <dd>Enter your rule’s test code in this pane and click the <strong>Run button</strong> to verify that the rule runs as intended. A <strong>matches panel</strong> appears after Semgrep is run to display matches and tests.</dd>
    <dt>Top menu</dt>
    <dd>Save, share, and add your rule to the Rule Board through this menu.</dd>
</dl>

To **resize the panes**, position your mouse over the borders and click-drag to the desired width. The library pane can be completely hidden to give more space for the editing panes.

Semgrep Registry rules are initially grouped by **directory**. Most of these directories correspond to languages. The Library can also be grouped by **rulesets**, which are rules arranged by category, such as security, best practices, and frameworks.

To group by ruleset, right click on the empty space on the **official registry** entry and select **Group by ruleset**.

> ![editor-groupbyruleset.png](../img/editor-groupbyruleset.png)

Reviewer 2: For this screenshot and a few of the others I think it would help to add a light border, since the screenshot’s background and the page’s background are both white (in non-dark mode). I’m not sure the easiest way to do this. It could be in the screenshot directly, or you could replace the Markdown image with the html tag for the image (`<img alt="editor-groupbyruleset.png" src= bla bla bla`) and add an inline style directly or add a CSS class that then attaches something like `border: 3px solid #eee;` or something like that.

Author: I agree! And I think some of the screenshots should also be smaller... for now I'll try to add a css value for width.

To create a rule, click **Create rule** on the splash page or the **(+) sign** next to the Library label. The Editor provides the minimum fields to create a rule:

- `id`
- `pattern`
- `message`
- `languages`
- `severity`

Refer to [Rule syntax](https://semgrep.dev/docs/writing-rules/rule-syntax/) for all possible fields and values to create a rule.

### Jumpstart rule writing using existing rules

Another method of creating rules is by **forking/copying** from existing rules found in Semgrep Registry. For example, Semgrep’s Java `crypto` ruleset prohibits the use of weak hashing algorithms `SHA-1` and `MD5`. An organization may want to further limit the use of other hash functions as part of their standards or security compliance. The next steps illustrate forking through copying an existing `use-of-sha1` rule and changing the rule to forbid MD2 hashes.

1. Enter related terms into the search bar to find potential rules. In our example, this is `SHA1.`
![Screenshot of Library pane with SHA1 filter](../img/editor-SHA1.png)

2. Registry rules cannot be edited directly. They can be **forked** by **right-clicking on their entry** in the Library and selecting **Fork rule.**

> ![Screenshot of forking menu](../img/editor-forking.png)

Reviewer 2: This screenshot could also use a soft border and maybe some margin or padding around it, since there is written text very close to it and the screenshot itself includes text as well.

Author: Will do. (And also noting for future exquisiteness when we get there.)

3. The rule is copied to your organization’s rules.
4. Enter edits to finalize your rule. In this example, the patterns are changed to find matches for [MD2](https://docs.oracle.com/javase/9/docs/specs/security/standard-names.html#messagedigest-algorithms) and the severity is increased from `WARNING` to `ERROR`.
5. Enter updates to test cases.
6. Click **Run** to validate your rule.
7. Click **Save** to save your rule. The following rule displays the end result.

<iframe title="Prevent use of MD2" src="https://semgrep.dev/embed/editor?snippet=s-santillan:use-of-md2" width="100%" height="432px" frameBorder="0"></iframe>

### Debugging syntax issues

> The editor will not save, share, or run if the YAML syntax is **unparseable**. The editor supports error handling, such as for the following cases:

Reviewer: ```suggestion
The editor will not save, share, or run if the YAML syntax is **unparseable**. The editor supports error handling, for the following cases:
```

If no schema or spelling errors are detected, check for indentation or newline issues. You can also get rule-writing help through the [community Slack](https://r2c.dev/slack).

## Evaluating a rule’s performance

Evaluating a rule through testing ensures that it meets your standards for precision and speed, thus **reducing false positives**. Semgrep uses certain annotations in comments to enable users to test which blocks of code are intended to match with a rule, and which blocks of code should not match.

In the following example, run the rule and click on the three-dot menu to view the result of the tests.

<iframe title="Testing example" src="https://semgrep.dev/embed/editor?snippet=s-santillan:snippet=7n1W" width="100%" height="432px" frameBorder="0"></iframe>

To **test a rule**:

1. Create at least one **true positive**: a code sample intended to match the rule.
2. Above this potential match, create a comment, followed by a space (` `), followed by `ruleid:RULE_ID` which specifies the rule that should match. In the preceding example, this is `//ruleid:hardcoded-conditional`.
3. Create at least one **true negative**: a code sample intended not to match the rule.

> 4. Above this non-match, create a comment followed by a space ( ), followed by `ok:RULE_ID`.

Reviewer: ```suggestion
4. Above this non-match, create a comment followed by a space (` `), followed by `ok:RULE_ID`.
```

Author: Have addressed this by writing out the example in full.

## Renaming and saving a rule

- To rename a rule, enter the new name in the YAML editor’s `id` field.
- To save a rule, enter **Ctrl+S** or click the **Save** button.

## Sharing a rule and setting a rule’s visibility

Upon saving, a rule’s visibility is **unlisted** by default. This rule can be shared with anyone through an identifier, even to non-Semgrep App users.

A rule can be saved as a **private rule**, which is visible only to members within an organization. You can still share a private rule, but only members of the organization will see it. Private rules are a **Team/Enterprise tier feature**.

- To set a rule’s visibility to private, click **Share > Private > Save change**.
- To share a private or unlisted rule, click **Share** and copy the **URL link**.

## Deleting a rule

To delete a rule, right click on its entry in the Library pane and select **Delete rule**. Deleting a rule is permanent. If the rule was previously added to the Rule Board, it is removed upon deletion.

## Setting code standards with the Rule Board

> Adding a rule to the Rule Board institutionalizes the rule to be applied to all projects scanned by Semgrep App. The rule can be added to either the **Audit** or **Blocker** columns. Rules in the Blocker column will prevent PRs from being merged until the finding is resolved.

Reviewer: ```suggestion
Adding a rule to the Rule Board applies the rule across all projects scanned by Semgrep App. The rule can be added to either the **Audit** or **Blocker** columns. Rules in the Blocker column will prevent PRs from being merged until the finding is resolved.
```

To add a rule to the Rule Board:

1. Click **Add to Rule Board**.
2. Select which column (Audit or Blocker) to add the rule to, depending on the rule’s relevance.
3. The rule appears on your Rule Board, under the corresponding column.

## Contributing to the open-source Semgrep Registry

> r2c, the maintainers of Semgrep, welcome contributions to the Registry. This is achieved by creating a pull request (PR) either on the repository itself or through the Editor. Contributions to the Registry can be discovered by anyone through the [Registry front-end](https://semgrep.dev/explore) and Editor.

Reviewer 2: ```suggestion
r2c, the maintainers of Semgrep, welcome contributions to the Registry. This is achieved by creating a pull request (PR) either on the repository itself or through the Editor. Contributions to the Registry can be discovered by anyone through the [Registry](https://semgrep.dev/explore) and Editor.
```

I think “front-end” doesn’t need to be specified.
